test/utils.h

```diff
@@ -441,6 +442,11 @@ class NVFuserTest : public ::testing::Test {
     // random seed. Otherwise, use zero. If a test fails, this seed will be
     // printed.
     std::srand(getCRandomSeed());
+
+    // For the NVFuser tests, we generally want to test Fuser's matmul
+    // scheduling and lowering. Thus disabling execution via expression
+    // evaluator.
+    DisableOptionsGuard::getCurOptions().set(DisableOption::MatmulExprEval);
```

Comment:
This creates a divergence between running nvFuser through bin/nvfuser_tests and through python_tests or just an arbitrary Python script. 

If reviewers agree to make MatmulExprEval opt out, I'd only disable it in test fixtures that we want it disabled. If there turn out to be too many fixtures to disable, that's probably an indicator that it should be opt-in instead :)

Reply:
vote on @jacobhinkle 's suggestion. Let's use a guard and not leave a corrupted flag after tests.

Reply:
Could we instead do this just for the Matmul test subclass?

Reply:
> Create a DisableGuard object that will be deconstructed in Teardown(). You can use unique_ptr to do this safely. That way it will automatically restore settings.

That's magic I'm not aware of! For my education, how do you make Setup() create an object that is automatically destructed in TearDown()? (I only know the manual way: https://github.com/NVIDIA/Fuser/blob/195a3f75d728739d8b386520e00540d59766ce3e/test/test_gpu_transpose.cpp#L45-L57)

Reply:
Never mind -- I figured out what you mean and created https://github.com/NVIDIA/Fuser/pull/1940 to simplify. unique_ptr is not necessary and we should do things [in the constructor](https://github.com/google/googletest/blob/main/docs/faq.md#should-i-use-the-constructordestructor-of-the-test-fixture-or-setupteardown-ctorvssetup) whenever possible.

Reply:
^^^ Thanks I actually thought `setup/teardown` is the preferred way :exploding_head:

Reply:
> That's magic I'm not aware of! For my education, how do you make Setup() create an object that is automatically destructed in TearDown()? (I only know the manual way:

Sorry for the slow response. Yes, I meant the manual way of putting `options_guard_ptr_ = nullptr;` in `Teardown()`;

Reply:
> would each TEST be considered to _always_ start from a fresh environment?

I believe so, yes. Are you thinking that there might be a case where we have two nested guards and we lose the middle state when we destroy the inner-most guard? In that case, the inner-most guard will be restoring the state it was initialized with, which should be whatever it was at initialization. This can only go wrong if an outer guard modifies its state during the lifetime of an inner guard, which will result in losing the updates when the inner guard is destroyed. Confusing enough?

Reply:
> would each TEST be considered to always start from a fresh environment?

According to [this section](https://github.com/google/googletest/blob/main/docs/primer.md#test-fixtures-using-the-same-data-configuration-for-multiple-tests-same-data-multiple-tests):

> For each test defined with TEST_F(), GoogleTest will create a fresh test fixture at runtime, immediately initialize it via SetUp(), run the test, clean up by calling TearDown(), and then delete the test fixture.

so constructor should be as fresh as setup.

Reply:
^^^ Thanks. I think I was just paranoid.

The doc JIngyue reference does seem to read like each test is an independent piece and ctor/setup considered as the singularity point where everything starts. So I shouldn't worry about restoring anything beyond that point..

Reply:
Yes, thank you @wujingyue for pointing to that doc. I had not made any assumptions on when the test object was created but it seems like it's actually preferred to use the ctor/dtor unless there is a good reason not to. That should work just fine with the guard. We can just default-initialize the options guard then set the option in the ctor using `DisableOptionsGuard::getCurOptions().set(DisableOption::MatmulExprEval);`